File: src/main/java/cc/sukazyo/cono/morny/bot/api/InputCommand.java
package cc.sukazyo.cono.morny.bot.api;

import cc.sukazyo.cono.morny.util.StringUtils;

import javax.annotation.Nonnull;
import javax.annotation.Nullable;
import java.util.Arrays;

public class InputCommand {
	
	private final String target;
	private final String command;
	private final String[] args;
	
	private InputCommand (@Nullable String target, @Nonnull String command, @Nonnull String[] args) {
		this.target = target;
		this.command = command;
		this.args = args;
	}
	
	public InputCommand (@Nonnull String[] inputArray) {
		this(parseInputArray(inputArray));
	}
	
	public InputCommand (@Nonnull String input) {
		this(StringUtils.formatCommand(input));
	}
	
	public InputCommand (@Nonnull InputCommand source) {
		this(source.target, source.command, source.args);
	}
	
	public static InputCommand parseInputArray (@Nonnull String[] inputArray) {
		final String[] cx = inputArray[0].split("@", 2);
		final String[] args = new String[inputArray.length-1];
		System.arraycopy(inputArray, 1, args, 0, inputArray.length - 1);
		return new InputCommand(cx.length == 1 ? null : cx[1], cx[0], args);
	}
	
	@Nullable
	public String getTarget () {
		return target;
	}
	
	@Nonnull
	public String getCommand () {
		return command;
	}
	
	@Nonnull
	public String[] getArgs () {
		return args;
	}
	
	public boolean hasArgs () {
		return args.length != 0;
	}
	
	@Override
	@Nonnull
	public String toString() {
		return String.format("{{%s}@{%s}#{%s}}", command, target, Arrays.toString(args));
	}
	
}

File: src/main/scala/cc/sukazyo/cono/morny/util/tgapi/InputCommand.scala
package cc.sukazyo.cono.morny.util.tgapi

import cc.sukazyo.cono.morny.util.UniversalCommand

class InputCommand private (
		val target: String|Null,
		val command: String,
		val args: Array[String]
) {
	
	override def toString: String =
		s"{{$command}@{$target}#{${args.mkString}}"
	
}

object InputCommand {
	
	def apply (input: Array[String]): InputCommand = {
		val _ex = if input.nonEmpty then input(0).split("@", 2) else Array.empty[String]
		val _args = input drop 1
		new InputCommand(
			if _ex.length > 1 then _ex(1) else null,
			if _ex.nonEmpty then _ex(0) else "",
			_args
		)
	}
	
	//noinspection NoTailRecursionAnnotation
	def apply (input: String): InputCommand =
		InputCommand(UniversalCommand.Lossy(input))
	
}
